File: C++/Step16-Queue and Deque/1021.cpp
//
//  1021.cpp
//  Step16-Queue and Deque
//
//  Created by 김예빈 on 2019. 9. 24..
//  Copyright © 2019년 김예빈. All rights reserved.
//

#include <iostream>
#include <deque>
using namespace std;

int main(int argc, const char * argv[]) {
    cin.tie(NULL);
    ios::sync_with_stdio(false);
    
    int n, m, input, idx, cnt = 0, left, right;
    deque<int> dq;
    deque<int>::iterator it;
    
    cin >> n >> m;
    
    for(int i = 1; i <= n; i++)
        dq.push_back(i);
    
    for(int i = 0; i < m; i++) {
        cin >> input;
        
        if(dq.front() == input)
            dq.pop_front();
        else {
            idx = 1;
            
            for(it = dq.begin(); it < dq.end(); it++) {
                if(*it == input) break;
                
                idx++;
            }
            
            left = idx - 1;
            right = dq.size() - idx + 1;
            
            if(left >= right) {
                for(int j = 0; j < right; j++) {
                    dq.push_front(dq.back());
                    dq.pop_back();
                }
                
                cnt += right;
                dq.pop_front();
                
            } else if(left <= right) {
                for(int j = 0; j < left; j++) {
                    dq.push_back(dq.front());
                    dq.pop_front();
                }
                
                cnt += left;
                dq.pop_front();
            }
        }
    }
    
    cout << cnt << "\n";
    
    return 0;
}

File: BaekJoon Online Judge/Swift/Step19-Queues, Deques/1021.swift
//
//  1021.swift
//  Step18-Queues and Deques
//
//  Created by Yebin Kim on 2019/12/28.
//  Copyright © 2019 김예빈. All rights reserved.
//

import Foundation

public struct Deque<T> {
    private var deque = Array<T>()
    
    public mutating func push_front(_ element: T) {
        deque.insert(element, at: 0)
    }
    
    public mutating func push_back(_ element: T) {
        deque.append(element)
    }
    
    public mutating func pop_front() {
        deque.removeFirst()
    }
    
    public mutating func pop_back() {
        deque.removeLast()
    }
    
    public var size: Int {
        return deque.count
    }
    
    public var front: T? {
        return deque.first
    }
    
    public var back: T? {
        return deque.last
    }
}

extension Deque: Sequence {
    public func makeIterator() -> AnyIterator<T> {
        return AnyIterator(IndexingIterator(_elements: deque.lazy))
    }
}


let input = readLine()!.split(separator: " ").map { Int($0)! }
let N = input[0]
let M = input[1]
var dq = Deque<Int>()
var answer = 0

for e in stride(from: 1, to: N + 1, by: 1) {
    dq.push_back(e)
}

let num = readLine()!.split(separator: " ").map { Int($0)! }

for e in num {
    if e == dq.front {
        dq.pop_front()
    } else {
        var idx = 1
        
        for it in dq {
            if it == e { break }
            
            idx += 1
        }
        
        let left = idx - 1;
        let right = dq.size - idx + 1;
        
        if left >= right {
            for _ in 0..<right {
                dq.push_front(dq.back!);
                dq.pop_back();
            }
            
            answer += right;
            dq.pop_front();
            
        } else if left <= right {
            for _ in 0..<left {
                dq.push_back(dq.front!);
                dq.pop_front();
            }
            
            answer += left;
            dq.pop_front();
        }
    }
}

print(answer)
